docs/tutorials/data_provider.md

```diff
@@ -61,3 +61,100 @@ impl AdditiveIdentity {
     }
 }
 ```
+
+## Caching Data Provider
+
+It is not generally required to implement a cache, especially if using the BakedDataProvider. However, if you wish to implement a cache in your data pipeline, you can do so like this:
+
+```rust
+use icu_provider::hello_world::HelloWorldFormatter;
+use icu_provider::prelude::*;
+use icu::locid::locale;
+use lru::LruCache;
+use std::convert::TryInto;
+use std::sync::Mutex;
+use yoke::trait_hack::YokeTraitHack;
+use yoke::Yokeable;
+use zerofrom::ZeroFrom;
+
+pub struct MyLruDataCache<P> {
+    cache: Mutex<LruCache<(DataKey, DataLocale), AnyResponse>>,
+    provider: P,
+}
+
+impl<M, P> DataProvider<M> for MyLruDataCache<P>
+where
+    M: KeyedDataMarker + 'static,
+    M::Yokeable: ZeroFrom<'static, M::Yokeable>,
+    M::Yokeable: icu_provider::MaybeSendSync,
+    for<'a> YokeTraitHack<<M::Yokeable as Yokeable<'a>>::Output>: Clone,
+    P: DataProvider<M>,
+{
+    fn load(&self, req: DataRequest) -> Result<DataResponse<M>, DataError> {
+        let cache_key = (M::KEY, req.locale.clone());
+        {
+            let mut cache = self.cache.lock().unwrap();
+            if let Some(any_resp) = cache.get(&cache_key) {
+                return any_resp.clone_downcast();
+            }
+        }
+        let computed_resp: DataResponse<M> = self.provider.load(req)?;
+        let computed_any_resp: AnyResponse = computed_resp.wrap_into_any_response();
+        {
+            let mut cache = self.cache.lock().unwrap();
+            let any_resp = cache.get_or_insert(cache_key, || computed_any_resp);
+            return any_resp.clone_downcast();
+        }
```

Comment:
This scope is pointless

Reply:
It's good practice to put mutexes inside explicit scopes... even if at the end of a function

Reply:
Says who? I think pointless scopes add confusion

Reply:
I added an additional comment at the top of the scope.

Reply:
No? 

Note this can be a one-liner, we can delay this debate to another time:

```
self.cache.lock().unwrap().get_or_insert(cache_key, || computed_any_resp).clone_downcast();

Reply:
I still prefer the explicitness of when locks are obtained and released, given that this is a common source of bugs in concurrent code. But I'll do the one-liner.